Allow cron.present to change a timespec from non-special to special

asomers opened the pull request:
### What does this PR do?

Previously when changing a cron.present job from using a normal time
specification to special or vice-versa, it would duplicate the entry in
the crontab.

### What issues does this PR fix or reference?
Fixes: #60997

### Previous Behavior
cron.present would duplicate cron jobs when changing timespec from non-special to special

### New Behavior
cron.present will delete the old non-special cronjob when creating a special one with the same identifier

### Merge requirements satisfied?
- [N/A] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

asomers committed:
Allow cron.present to change a timespec from non-special to special

Previously when changing a cron.present job from using a normal time
specification to special or vice-versa, it would duplicate the entry in
the crontab.

asomers commented:
The test failure looks unrelated to this PR.

asomers commented:
ping @twangboy would you be able to review this PR sometime?

asomers commented:
rebased @twangboy

garethgreenaway commented:
@asomers Looks good.  Can you please add a changelog to this PR?  Thanks!

twangboy reviewed (dismissed):
Needs a changelog

asomers commented:
Rebased and changelog added, @twangboy @garethgreenaway

garethgreenaway commented:
re-run all

Ch3LL commented on salt/states/cron.py:
```diff
@@ -355,6 +355,9 @@ def present(
         return ret
 
     if special is None:
+        antidata = __salt__["cron.rm_special"](
```
I would prefer the logic be in the `set_job` and `set_special` functions. Those functions do some logic to see what is currently set in the crontab and see if there is a difference. We should be able to add the logic in those functions to determine if there needs to be an update instead of removing it first.

asomers replied:
@Ch3LL I made your requested change.  And it turned out that only one of the existing modules/cron tests needed fixing.

asomers committed:
Fix the assertions in test_set_job

The original assertion used "mock_method.call_args.assert_called_with",
which does nothing at all.  Correct usage is either
"mock_method.call_args == ..." or "mock_method.assert_called_with(...)"

asomers committed:
Move logic from states/cron into modules/cron

This allows the cron.set_job and cron.set_special commands to change
cron entries from special to non-special or vise-versa.  It also removes
all changes to states/cron from this PR's original commit, while
retaining the behavioral changes to the cron.present state
